components/dashboard/src/insights/download/download-sessions.ts

```diff
@@ -0,0 +1,241 @@
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import {
+    ListWorkspaceSessionsRequest,
+    WorkspaceSession,
+    WorkspaceSpec_WorkspaceType,
+} from "@gitpod/public-api/lib/gitpod/v1/workspace_pb";
+import { organizationClient, workspaceClient } from "../../service/public-api";
+import dayjs from "dayjs";
+import { useQuery, useQueryClient } from "@tanstack/react-query";
+import { useCallback } from "react";
+import { noPersistence } from "../../data/setup";
+import { Timestamp } from "@bufbuild/protobuf";
+import type { OrganizationMember } from "@gitpod/public-api/lib/gitpod/v1/organization_pb";
+
+const pageSize = 100;
+const maxPages = 100; // safety limit if something goes wrong with pagination
+
+type GetAllWorkspaceSessionsArgs = Pick<ListWorkspaceSessionsRequest, "to" | "from" | "organizationId"> & {
+    signal?: AbortSignal;
+    onProgress?: (percentage: number) => void;
+};
+export const getAllWorkspaceSessions = async ({
+    from,
+    to,
+    signal,
+    organizationId,
+    onProgress,
+}: GetAllWorkspaceSessionsArgs): Promise<WorkspaceSession[]> => {
+    const records: WorkspaceSession[] = [];
+    let page = 0;
+    while (!signal?.aborted && page < maxPages) {
+        const response = await workspaceClient.listWorkspaceSessions(
+            {
+                organizationId,
+                from,
+                to,
+                pagination: {
+                    page,
+                    pageSize,
+                },
+            },
+            {
+                signal,
+            },
+        );
+        if (response.workspaceSessions.length === 0) {
+            break;
+        }
+
+        records.push(...response.workspaceSessions);
+        onProgress && onProgress(page);
+
+        page = page + 1;
+    }
+
+    return records;
+};
+
+type Args = Pick<ListWorkspaceSessionsRequest, "organizationId" | "from" | "to"> & {
+    organizationName: string;
+    signal?: AbortSignal;
+    onProgress?: (percentage: number) => void;
+};
+
+export type DownloadUsageCSVResponse = {
+    blob: Blob | null;
+    filename: string;
+    count: number;
+};
+
+const downloadUsageCSV = async ({
+    organizationId,
+    from,
+    to,
+    organizationName,
+    signal,
+    onProgress,
+}: Args): Promise<DownloadUsageCSVResponse> => {
+    const start = dayjs(from?.toDate()).format("YYYYMMDD");
+    const end = dayjs(to?.toDate()).format("YYYYMMDD");
+    const filename = `gitpod-usage-${organizationName}-${start}-${end}.csv`;
+
+    const records = await getAllWorkspaceSessions({
+        organizationId,
+        from,
+        to,
+        signal,
+        onProgress,
+    });
+
+    if (records.length === 0) {
+        return {
+            blob: null,
+            filename,
+            count: 0,
+        };
+    }
+
+    const orgMembers = await organizationClient.listOrganizationMembers({
```

Comment:
This is not ideal, but works as long as we don't introduce pagination into the list org members method. It would be awesome if we could do one of the following:
- introduce a `OrganizationService.GetMember` method to not have to list hundreds of members who possibly did not even create workspaces in the specified period
- add the owner meta on the response of `WorkspaceService.ListWorkspaceSessions`

Reply:
:ok:  IMO ok for now, but definitely worth a comment.

Reply:
Just checked, it seems we only use the `fullName` field... not sure whether it belongs into Workspace, but we can add it to the `WorkspaceSession` shape for sure.

Reply:
That sounds fair, like having an `owner` field on the session with those limited details

Reply:
Nice, this cleaned up a lot of the frontend code: [`f77d7f8` (#20437)](https://github.com/gitpod-io/gitpod/pull/20437/commits/f77d7f858e71f4bbf053aae36339fb6322d43467).